### ISSUE TITLE
`BaseAnalysis`: Investigate why `join` over alternatives for non-definite AD target does not work

### ISSUE THREAD

**Issue by michael-schwarz** [goblint/analyzer#1465]
For #1458 I also attempted to replace this snippet 

https://github.com/goblint/analyzer/blob/5cd8650a2872341152f48a7d1cd20ac9fa0de994/src/analyses/base.ml#L1735-L1743

with

~~~OCaml
AD.fold (fun addr acc -> D.join (update_one addr st) acc) lval (D.bot ())
~~~

but that lead to some strange test failures where I really have no idea what's going on:

```
3 test(s) failed: ["04/35 trylock_rc", "04/36 trylock_nr", "41/05 more"]
```

This would be the cleaner solution, so we should investigate it.

_Originally posted by @michael-schwarz in https://github.com/goblint/analyzer/pull/1458#discussion_r1599661735_

**Comment by michael-schwarz** [comment]
That's odd indeed. I'm fine with the current improvement then because it fixes the issue and doesn't break anything else.
The naive implementation failing some tests deserves an issue though. I had a quick tracing look and somehow there seemed to be `CPA` maps with many variables missing or `Uninitialized` (i.e. superbot) values, so something odd happens. Maybe when the address set is somehow empty, the new `D.bot ()` comes into play?

_Originally posted by @sim642 in https://github.com/goblint/analyzer/pull/1458#discussion_r1599943567_

**Comment by karoliineh** [comment]
The tests pass when we replace it with:
```c
if AD.is_empty lval then
  st
else
  AD.fold (fun addr acc -> D.join (update_one addr st) acc) lval (D.bot ())
```

1. In "04/35 trylock_rc", and "04/36 trylock_nr", `lval` becomes empty because of the call `time (NULL)`, where in LibFuns we try to write to the argument (`("time", unknown [drop "arg" [w]]);`) that is then a `NullPointer`, which is removed by `collect_invalidate`.
2. In "41/05 more" the same happens because of the call `strtol("bla", 0, 8)`, where we try to write to the second argument `("strtol", unknown [drop "nptr" [r]; drop "endptr" [w]; drop "base" []]);` that is also a `NullPointer`.

I tried to catch all cases where `lval` might be empty with `assert (not @@ AD.is_empty lval);`. Then one more test ("05/21 race-null-type") crashes, but that also contains a call to `time (NULL)`.
